pkg/kvstore/allocator.go

```diff
@@ -0,0 +1,502 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kvstore
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/nodeaddress"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	// maxAllocAttempts is the number of attempted allocation requests
+	// performed before failing.
+	maxAllocAttempts = 16
+
+	gcInterval = time.Duration(10) * time.Minute
+)
+
+// ID is the allocated identifier which maps to a key
+type ID uint64
+
+// String returns the string representation of an allocated ID
+func (i ID) String() string {
+	return strconv.FormatUint(uint64(i), 10)
+}
+
+// IDMap provides mapping from ID to key
+type IDMap map[ID]string
+
+// Allocator is a distributed ID allocator backed by a KVstore. It maps
+// arbitrary keys to identifiers. Multiple users on different cluster nodes can
+// in parallel request the ID for keys and are guaranteed to retrieve the same
+// ID for an identical key.
+//
+//
+// Key hierarchy
+//   slave keys:
+//    - basePath/value/key1/node1 => 1001
+//    - basePath/value/key1/node2 => 1001
+//    - basePath/value/key2/node1 => 1002
+//    - basePath/value/key2/node2 => 1002
+//
+//   master key:
+//    - basePath/id/1001 => key1
+//    - basePath/id/1002 => key2
+type Allocator struct {
+	// Events is a channel which will receive AllocatorEvent as IDs are
+	// added, modified or removed from the allocator
+	Events AllocatorEventChan
```

Comment:
The identity code which should trigger a policy recalc when a new identity has been allocated in the cluster.